hazelcast-client-new/src/main/java/com/hazelcast/client/cache/impl/AbstractClientInternalCacheProxy.java

```diff
@@ -89,22 +113,68 @@
 
     private final AtomicInteger completionIdCounter = new AtomicInteger();
 
-    protected AbstractClientInternalCacheProxy(CacheConfig cacheConfig, ClientContext clientContext) {
+    protected AbstractClientInternalCacheProxy(CacheConfig cacheConfig, ClientContext clientContext,
+                                               HazelcastClientCacheManager cacheManager) {
         super(cacheConfig, clientContext);
+        this.cacheManager = cacheManager;
+        nearCacheManager = clientContext.getNearCacheManager();
         asyncListenerRegistrations = new ConcurrentHashMap<CacheEntryListenerConfiguration, String>();
         syncListenerRegistrations = new ConcurrentHashMap<CacheEntryListenerConfiguration, String>();
         syncLocks = new ConcurrentHashMap<Integer, CountDownLatch>();
 
         NearCacheConfig nearCacheConfig = cacheConfig.getNearCacheConfig();
         if (nearCacheConfig != null) {
-            nearCache = new ClientHeapNearCache<Data>(nameWithPrefix, clientContext, nearCacheConfig);
             cacheOnUpdate = nearCacheConfig.getLocalUpdatePolicy() == NearCacheConfig.LocalUpdatePolicy.CACHE;
+            NearCacheContext nearCacheContext =
+                    new NearCacheContext(clientContext.getSerializationService(),
+                            createNearCacheExecutor(clientContext.getExecutionService()));
+            nearCache = nearCacheManager
+                    .getOrCreateNearCache(nameWithPrefix, nearCacheConfig, nearCacheContext);
+            registerInvalidationListener();
         } else {
             nearCache = null;
             cacheOnUpdate = false;
         }
     }
 
+    protected class ClientNearCacheExecutor implements Executor {
```

Comment:
instead of creating a wrapper class you could've just extend `ClientExecutionService` with `Executor`
it has the the method (`void execute(Runnable command)`) already